**Ex2 - Getting and Knowing your Data**
This time we are going to pull data directly from the internet. Special thanks to: https://github.com/justmarkham for sharing the dataset and materials.

### **Step 1. Import the necessary libraries**

In [1]:
import pandas as pd
import numpy as np

### **Step 2. Import the dataset from this**

In [119]:
data = pd.read_csv('https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv', sep = '\t')

### **Step 3. Assign it to a variable called chipo**

In [120]:
chipo = data

### **Step 4. See the first 10 entries**

In [8]:
chipo.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
6                                                NaN     $1.69   
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75   
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25   
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25

### **Step 5. What is the number of observations in the dataset?**

In [10]:
chipo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


There is a total of 4,622 observations, although choice description only has 3,376 non-null values

### **Step 6. What is the number of columns in the dataset?**

In [12]:
len(chipo.columns)

5

In [14]:
chipo.shape

(4622, 5)

In [17]:
chipo.shape[1]

5

There are 5 columns

### **Step 7. Print the name of all the columns**

In [18]:
chipo.columns

Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')

In [19]:
chipo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


### **Step 8. How is the dataset indexed?**

In [20]:
chipo.index

RangeIndex(start=0, stop=4622, step=1)

The Index of the dataset is a RangeIndex from 0 to 4,622 in steps of 1

<div class="alert alert-block alert-warning">
    For this question, I misinterpreted it as the most ordered item in a single order, when they were looking for the most ordered item across all order

### **Step 9. Which was the most-ordered item?**

In [23]:
chipo[chipo['quantity'] == max(chipo['quantity'])]

order_id  quantity                     item_name choice_description  \
3598      1443        15  Chips and Fresh Tomato Salsa                NaN   

     item_price  
3598    $44.25

The Most ordered item is 'Chips and Fresh Tomato Salsa'

#### **The correction**

In [96]:
chipo.groupby(['item_name']).agg({'quantity':'sum'}).reset_index().sort_values('quantity', ascending = False).head(1)

item_name  quantity
17  Chicken Bowl       761

In [109]:
chipo.groupby(['item_name']).agg({'quantity':'sum'}).reset_index().max()
#so the reason that Veffie Soft Tacos is showing up is because max() is looking at the max index and giving the item_name, then it is looking at the max quantity and giving the max there.

item_name    Veggie Soft Tacos
quantity                   761
dtype: object

In [115]:
#another way to do the same thing
grouped = chipo.groupby(['item_name']).agg({'quantity':'sum'})

In [114]:
grouped[grouped['quantity'] == max(grouped['quantity'])]

quantity
item_name             
Chicken Bowl       761

<div class="alert alert-block alert-warning">
    So since I misunderstood the above question, then I misunderstood this question also, the answer for Step 10. is 761

### **Step 10. For the most-ordered item, how many items were ordered?**

Chips and Fresh Tomato Salsa had 15 orders

<div class="alert alert-block alert-danger">
<b>Step 11. What was the most ordered item in the choice_description column?

I am not sure how to handle this becuase the choice_description column is a string that contains lists within lists

looking at sample solution, I thought I would have to parse through the choice descriptions, but instead they just used the whole value of the string, which seems a little disingenuous
because a choice_description could be '[Tomatoes, [Diet Coke, Eggs]]' and ['Tomatoes'] and here tomatoes would only have 1 count, even though it appears in both choice descriptions.

In [116]:
c = chipo.groupby('choice_description').sum()
c = c.sort_values(['quantity'], ascending=False)
c.head(1)
# Diet Coke 159

order_id  quantity  item_price
choice_description                                
[Diet Coke]           123455       159      326.71

### **Step 12. How many items were ordered in total?**

In [44]:
chipo.head()

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98

In [45]:
chipo['quantity'].sum()

4972

The total amount of items ordered was 4972

<div class="alert alert-block alert-warning">
    The sample solution for Step 13

In [121]:
dollarizer = lambda x: float(x[1:-1])
chipo.item_price = chipo.item_price.apply(dollarizer)

So they are making use of the fact that the '$' only appears at the first place in each element in the item_price column, so they are making a variable x, be everything after the dollar sign
and then converting that to a float, then using .apply() to apply the lambda function, very interesting

### **Step 13. Turn the item price into a float**

In [53]:
chipo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


In [72]:
chipo['item_price'] = chipo['item_price'].str.replace('$',"").astype(float)

C:\Users\cs.ABR\AppData\Local\Temp/ipykernel_13496/69540394.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  chipo['item_price'] = chipo['item_price'].str.replace('$',"").astype(float)


In [73]:
chipo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   order_id            4622 non-null   int64  
 1   quantity            4622 non-null   int64  
 2   item_name           4622 non-null   object 
 3   choice_description  3376 non-null   object 
 4   item_price          4622 non-null   float64
dtypes: float64(1), int64(2), object(2)
memory usage: 180.7+ KB


In [74]:
#they recommend using a lambda function, will look at sample solutions when done going through the exercises myself

<div class="alert alert-block alert-warning">
Revenue is the price * the quantity

### **Step 14. How much was the revenue for the period in the dataset?**

In [123]:
revenue = chipo['item_price'] * chipo['quantity']
revenue.sum()

39237.02

The revenue for the period is $39,237.02

<div class="alert alert-block alert-warning">
    There are multiple rows with the same order_id, So I should have used value_counts() and then sum()

### **Step 15. How many orders were made in the period?**

In [77]:
chipo['order_id'].count()

4622

There were 4,622 orders made in the period

Sample Solution for STEP 15

In [124]:
orders = chipo.order_id.value_counts().count()
orders

1834

<div class="alert alert-block alert-warning">
    Again, I did not properly calculated revenue, so must properly calculate it to get this question right

### **Step 16. What is the average revenue amount per order?**

In [126]:
revenue.mean()

8.489186499350943

The Average Revenue amount per order is $7.46

Looking at Sample Solution for STEP 16

In [127]:
chipo['revenue'] = chipo['quantity'] * chipo['item_price']
order_grouped = chipo.groupby(by=['order_id']).sum()
order_grouped.mean()['revenue']

21.394231188658654

### **Step 17. How many different items are sold?**

In [80]:
chipo['item_name'].nunique()

50

There are 50 different items sold.

## Overall, I learned to understand the dataset more before performing operations on it and to better think about the questions before I answer, If I did that, then I would have gotten All these question right